api/workload/v2/reader.go

```diff
@@ -0,0 +1,94 @@
+package workload
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/api/workload"
+)
+
+type streamReader struct {
+	Chan          chan *workload.X509SVIDResponse
+	logger        *logrus.Logger
+	streamManager *streamManager
+
+	// atomic stream
+	mu     *sync.RWMutex
+	stream *managedStream
+}
+
+func newStreamReader(ctx context.Context, logger *logrus.Logger, streamManager *streamManager) *streamReader {
+	r := &streamReader{
+		logger:        logger,
+		streamManager: streamManager,
+		mu:            new(sync.RWMutex),
+		Chan:          make(chan *workload.X509SVIDResponse),
+	}
+	r.start(ctx)
+	return r
+}
+
+func (c *streamReader) getStream() *managedStream {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.stream
+}
+
+func (c *streamReader) setStream(stream *managedStream) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.stream = stream
+}
+
+func (c *streamReader) start(ctx context.Context) {
+	c.logger.Debug("Starting reader.")
+	go func() {
+		defer c.logger.Debug("Shutting down reader")
+		defer c.closeStream()
+		defer close(c.Chan)
+
+		for {
+
+			select {
+			case stream, ok := <-c.streamManager.Chan:
+				if !ok {
+					return
+				}
+				c.setStream(stream)
+			case <-ctx.Done():
+				return
+			}
+
+			for {
+				resp, err := c.getStream().Recv()
+				if err != nil {
+					c.logger.WithError(err).Info("Stream reader failed.")
+					c.closeStream()
+					c.streamManager.Reconnect()
+					break
+				}
+				c.Chan <- resp
+			}
+		}
+	}()
+}
+
+func (c *streamReader) closeStream() {
+	if stream := c.getStream(); stream != nil {
+		if err := stream.Close(); err != nil {
+			c.logger.WithError(err).Info("Stream close failed.")
+		}
+		c.setStream(nil)
+	}
+}
+
+func (c *streamReader) Stop() {
+	c.closeStream()
```

Comment:
Hmm, this should be cancelling a subcontext I think. This is mixing ownership of the stream and causing you to need a mutex. Is this necessary because stream.Recv() doesn't take a context?

Once you get the stream over the channel the `start` goroutine should own the full lifecycle of the stream and you can drop the mutex (assuming `Stop` has a way of unblocking `Recv()` also I think you will need to be checking ctx.Done() in there as well.